fdbserver/VersionedBTree.actor.cpp

```diff
@@ -4836,10 +4871,46 @@ struct DecodeBoundaryVerifier {
 			        b->second.upper.printable().c_str());
 			return false;
 		}
+
+		if (!b->second.domainId.present()) {
+			ASSERT(!keyProvider || !keyProvider->enableEncryptionDomain());
+			ASSERT(!domainId.present());
+		} else {
+			ASSERT(keyProvider->enableEncryptionDomain());
+			if (b->second.domainId != domainId) {
+				fprintf(stderr,
+				        "Page encrypted with incorrect domain: %s %s, using %s, written %s\n",
+				        ::toString(id).c_str(),
+				        ::toString(v).c_str(),
+				        ::toString(domainId).c_str(),
+				        ::toString(b->second.domainId).c_str());
+				return false;
+			}
+			if (deterministicRandom()->random01() < domainPrefixScanProbability) {
```

Comment:
I think it would be better and valid to simplify all of this to just moveFirst() and moveLast() on the cursor to get the first and last keys, get their common prefix string with commonPrefixLength(), and make sure that prefix is the expected domain.  Then you don't need the sample as this is cheap to verify every time.

The above is valid to check the domain ID for this page's data so long as DeltaTree does not break its contract, so in effect what this iterative check is validating is that DeltaTree is not breaking its contract.  There are other checks for that including a randomized unit test, and a lot of other things that would go wrong if this were to occur.

Reply:
Just checking the first and last keys are not sufficient in case the default domain is involved. We could have a page containing only domain prefixes (and thus encrypted using the default domain), and then accidentally a key with non-prefix portion is in-place inserted to the page. The check here verifies such case doesn't happen. That said, I will change to do the check only for default domain encrypted pages.